pkg/cli/create_helm.go

```diff
@@ -320,35 +327,66 @@ func isVClusterDeployed(release *helm.Release) bool {
 		release.Secret.Labels["status"] == "deployed"
 }
 
-// TODO Delete after vCluster 0.19.x resp. the old config format is out of support.
-// migratelegacyHelmValues migrates the values of the current vCluster to the new config format.
-// Only returns a non-empty string if the passed in release is < v0.20.0.
-func migrateLegacyHelmValues(release *helm.Release) (string, error) {
-	if semver.Compare("v"+release.Chart.Metadata.Version, "v0.20.0-alpha.0") != -1 {
-		// No need to migrate new releases.
-		return "", nil
+func isLegacyVCluster(version string) bool {
+	if version == upgrade.DevelopmentVersion {
+		return false
 	}
+	return semver.Compare("v"+version, "v0.20.0-alpha.0") == -1
+}
+
+func isLegacyConfig(values []byte) bool {
+	cfg := legacyconfig.LegacyK0sAndK3s{}
+	if err := cfg.UnmarshalYAMLStrict(values); err != nil {
+		// Try to parse it as k8s/eks
+		cfg := legacyconfig.LegacyK8s{}
+		if err := cfg.UnmarshalYAMLStrict(values); err != nil {
+			return false
+		}
+	}
+	return true
+}
 
-	// If we are upgrading a vCluster < 0.20 the old k3s chart is the one without a prefix.
-	distro := strings.TrimPrefix(release.Chart.Metadata.Name, "vcluster-")
-	if distro == "vcluster" {
-		distro = "k3s"
+// helmValuesYAML returns the values yaml from the secret inside the Helm release.
+func helmValuesYAML(release *helm.Release) ([]byte, error) {
+	enc, ok := release.Secret.Data["release"]
+	if !ok {
+		return nil, fmt.Errorf("no helm release values")
 	}
 
-	cfg := release.Config
-	y, err := yaml.Marshal(cfg)
+	// The helm release is a gzipped base64-encoded blob
+	dec, err := base64.StdEncoding.DecodeString(string(enc))
 	if err != nil {
-		return "", err
+		return nil, err
 	}
-	migratedValues, err := legacyconfig.MigrateLegacyConfig(distro, string(y))
+
+	gz, err := gzip.NewReader(bytes.NewReader(dec))
 	if err != nil {
-		return "", err
+		return nil, err
 	}
+	defer gz.Close()
 
-	return migratedValues, nil
-}
+	releaseRaw, err := io.ReadAll(gz)
+	if err != nil {
+		return nil, err
+	}
+
+	helmChart := struct {
+		Chart struct {
+			Values map[string]interface{} `json:"values"`
+		} `json:"chart"`
+	}{}
 
-// TODO end
+	if err := yaml.Unmarshal(releaseRaw, &helmChart); err != nil {
+		return nil, err
+	}
```

Comment:
can we use an encoder/decoder here instead of reading everything into memory or not?